### PR TITLE
Add some ICEs

### PR DESCRIPTION
**Pull request by ritikmishra**
- rust-lang/rust#106238
- rust-lang/rust#107409
- rust-lang/rust#108399
- rust-lang/rust#108498
- rust-lang/rust#108683
- rust-lang/rust#108814
- rust-lang/rust#109020

### CODE REVIEW

**Comment by JohnTitor** [comment]
https://github.com/rust-lang/rust/pull/108339 doesn't seem like an ICE issue.

**Comment by ritikmishra** [comment]
my bad - it was supposed to be rust-lang/rust#108399

File: ices/106238.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// Trait alias causes compiler panic
+pub trait Cell<T: Add<T, Output=T>> = DoSome<T>;
+
+struct _Container<T> {
+    pub cells: dyn Cell<T>,
+}
+
+
+fn main() {}
+
```
Comment (JohnTitor) [inline ices/106238.rs:15]:
```suggestion

fn main() {}
```

File: ices/107409.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+fn reroot<'gc, T: Rootable<'gc>>(_t: T, _f: for<'a> fn(T::AsRoot<'a>)) {
+}
+
+fn test<'gc>(t: Gc<'gc, i32>) {
+    reroot(t, |_| ());
+}
+
+fn main() {}
```
Comment (JohnTitor) [inline ices/107409.rs:24]:
```suggestion
use std::marker::PhantomData as Boo;

struct Gc<'gc, T: 'gc>(Boo<fn(&'gc T) -> &'gc T>);

trait Rootable<'env> {
    type AsRoot<'r>: Rootable<'r> + 'r
    where
        'env: 'r;
}

impl<'env, T: Rootable<'env>> Rootable<'env> for Gc<'env, T> {
    type AsRoot<'r> = Gc<'r, T::AsRoot<'r>> where 'env: 'r;
}

impl<'env> Rootable<'env> for i32 {
    type AsRoot<'r> = i32 where 'env: 'r;
}

fn reroot<'gc, T: Rootable<'gc>>(_t: T, _f: for<'a> fn(T::AsRoot<'a>)) {}

fn test<'gc>(t: Gc<'gc, i32>) {
    reroot(t, |_| ());
}

fn main() {}
```

File: ices/108399.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    call(operation).await
+}
+
+async fn call<F>(mut f: F)
+where
+    for<'any> F: FnMut(&'any mut ()) -> FutNothing<'any>,
+{
+    f(&mut ()).await
+}
+
```
Comment (JohnTitor) [inline ices/108399.rs:21]:
```suggestion
#![feature(type_alias_impl_trait)]

use std::future::Future;

type FutNothing<'a> = impl 'a + Future<Output = ()>;

async fn operation(x: &mut ()) -> () {
    ()
}

async fn indirect() {
    call(operation).await
}

async fn call<F>(mut f: F)
where
    for<'any> F: FnMut(&'any mut ()) -> FutNothing<'any>,
{
    f(&mut ()).await
}

fn main() {}
```

File: ices/108814.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#![feature(non_lifetime_binders)]
+
+fn take(_: impl for<T> FnOnce(T) -> T) {}
+
+fn main() {
+    take(|x| x)
+}
+
```
Comment (JohnTitor) [inline ices/108814.rs:8]:
```suggestion
#![feature(non_lifetime_binders)]

fn take(_: impl for<T> FnOnce(T) -> T) {}

fn main() {
    take(|x| x)
}
```

File: ices/109020.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+where
+    F: EmplacableFn,
+{
+    phantom: PhantomData<F>,
+}
+
+fn main() {
+    with_emplacable_for(|_| {});
+}
+
```
Comment (JohnTitor) [inline ices/109020.rs:38]:
```suggestion
```

**Review by JohnTitor — APPROVED** [review]
Thanks a lot!
Overall LGTM, fixed style and removed one non-ICE issue.

Some tips for next time:
- Format code with rustfmt and add the main function if possible; this work is helpful when we move code to rust-lang/rust's regression tests.
- Check if the issue is already resolved, this can be checked by `cargo run`

**Pull request merged**